### ISSUE TITLE
Threat of overloading cloud converters when license key is expired

### ISSUE THREAD

**Issue by rosell-dk** [rosell-dk/webp-convert#38]
The following scenario worries me:

1. There is only cloud conversion available on the server
2. The user buys a key
3. The key expires
4. The script which uses WebPConvert keeps asking the cloud converter to convert the same files, because it keeps seing that those files needs conversion.
5. The cloud converter gets overloaded with requests. Not because it needs to do a lot of conversions, but because it gets a lot of requests. That these requests has big uploads with them may worsen things.

This may turn real bad. If for example you have a site with many big images on the home page and an invalid ewww license key, and a script, such as [webp-on-demand](https://github.com/rosell-dk/webp-on-demand) running, then each time the homepage will be visited, there will be made ie 40 uploads of each 100 MB to the ewww cloud service. And if that site has many visitors, it becomes a real problem.

Any ideas ?

**Comment by rosell-dk** [comment]
We could automatically throw an TooManyFailedConversionsTodayException, if the number of failed conversions on a day exceeds a limit, which can be configured.

**Comment by rosell-dk** [comment]
Or perhaps a TooManyFailedConversionsInARowException, as the problem usually won't fix itself by the next day.

**Comment by rosell-dk** [comment]
And in order to fix that exception, you'd have to call a static method for re-enabling the converter.

**Comment by rosell-dk** [comment]
What would be the best way to store the count?

**Comment by rosell-dk** [comment]
I guess perhaps we should run a quota check first, before uploading a file to Ewww. But I will write to Shane

**Comment by S1SYPHOS** [comment]
... that‘s what isValidKey() was for, no?

**Comment by nosilver4u** [comment]
There are a couple options that the EWWW IO plugin uses. I originally set it up to run a key validation before every single image, but that proved to be unnecessary.

So then I backed it off to validate only once per hour, and store the validation status in a transient. In WordPress, those are basically temporary "options" that have an expiration date attached to them. So use whatever "temporary storage" seems best to you for that.

Then, there are two other status messages to consider. As you're aware, normally the key verification returns "great", but if a key has run out of quota, then it will return "exceeded", and if the key is not valid (not found in the system), it returns "invalid".
For "exceeded" keys, the plugin will stop doing anything for 5 minutes, so that the user doesn't have to wait so long when they purchase more credits. It could go longer if you have a way for the user to manually refresh/reset the timer. In WordPress, visiting the EWWW IO settings page resets the timer, but I haven't raised the timeout beyond 5 minutes (yet).
For "invalid" keys, the key should be removed from the configuration and not used anymore.

**Comment by rosell-dk** [comment]
Thanks, Shane!

**Comment by rosell-dk** [comment]
Yes, it makes sense to remove invalid keys. Only, we cannot do that, as WebPConvert does not store the key, but it is passed to our library. But what the WebPConvert *can* do is to add keys which are detected invalid to a blacklist. Once on the blacklist, WebPConvert can reject attempts to convert with that key immediately.

The idea of temporarily locking down for 5 minutes after a "exceeded" response is also great. After the end of a lock-out, I suggest that the next call checks quota before trying to do a real conversion. However, it is to be expected that many applications utilizing WebPConvert will not inform the user of exceeded quota. [webp-on-demand](https://github.com/rosell-dk/webp-on-demand) is for example very much a set up and forget approach. Having websites knocking on the ewww api door every five minutes the next many years is maybe a bit too often, even if they only knock lightly (without uploading a file). We could for example increase it over time (ie increase to 10 minutes when quota has been exceeded for a month, increase further to 15 minutes, when it has been exceeded for 6 months)

With implementation of these two rules, I think we will protect the ewww cloud service quite well from being overloaded with convert requests with invalid or expired keys.

There is the question if we should we in a addition to above should also validate key (or quota) before each conversion. I think it is not necessary. The first rule above makes sure that invalid keys are only used once. It will not harm the Ewww cloud service to have to reject a single upload. Actually, it will just put more pressure to the service, if it has to respond to a key validation test once for every conversion. Almost same reasoning goes to quota requests, as quota only exceed rarely. 
I actually cannot see the idea of setting the key validation down to once every hour, rather than just dropping it all together. Chances are quite against that the key suddenly is invalid, and if it is, it will be detected, and blocked forever. Chances of sporadic quota validations turning out negative seems equally small. I gather that it will generally be so, that if it worked last time, it will probably still work.

**Comment by nosilver4u** [comment]
In regards to the last point about key validation every hour, perhaps with your library it will be less common. But for example, I give out free keys to folks that opt into the usage tracking within EWWW IO. If those keys are used up and not converted to paying keys within 6 months, they are removed, and become invalid.

I think it's important to clarify that while the verification endpoint will return an image or the "exceeded" message, it will never return an "invalid", which is why the EWWW IO plugin does periodic verification requests. I know that's probably a little weird, but the whole purpose of the verification endpoint is to allow easy/quick updates regarding the API key. It's nearly impossible to overwhelm the API with the /verify/ requests, so you don't have to worry about that.

**Comment by rosell-dk** [comment]
I'm a bit confused. I guess you meant the "convert" endpoint in your last paragraph? 

- Because In the documentation (https://ewww.io/api/), you write that an empty response to /verify/ indicates an invalid key. And my tests show the same. So it seems that we can rely on that.

**Comment by rosell-dk** [comment]
I think we will go with sending a /verify/ before each /convert/, then (and no other mechanisms). Mainly because it is quickly set up. The result is slower conversion, as each conversion requires two requests. So at some point, we can improve the implementation by adding blacklist and temporary lock-outs.

**Comment by rosell-dk** [comment]
Sending verification requests before each conversion request is inefficient. Reopening!

**Comment by rosell-dk** [comment]
So, the challenge here is that webp-convert is stateless. However, projects using webp-convert are not. These projects should have a chance to turn off the verification requests for each conversion and implement their own handling of this. 

To assist that kind of handling, this should throw more nuanced errors appropriate errors depending on the status of a convert call. Currently all responses that are JSON and has an "error" property results in an `InvalidApiKey` exception. But if we receive ie {"error":"invalid","t":"exceeded"}, that info - that the quota was exceeded - should be made available so the project can take appropriate action.

Further, it should be stipulated in the documentation that bypassing the verify request should only be done if one makes sure to catch that type of exception and reacts to it by disabling the api-key.

**Comment by nosilver4u** [comment]
In case it is helpful for handling different errors, the EWWW IO convert/optimize endpoint will now return an "invalid" response for bogus/expired keys. Previously, the API returned "exceeded" for any key that wasn't valid (and of course, for a key without any more credits available).

**Comment by nosilver4u** [comment]
Also, it would be useful for debugging on my end if this library used a different user-agent. I noticed in the code that it has this:

```
        // The 403 forbidden is avoided with this line.
        curl_setopt(
            $ch,
            CURLOPT_USERAGENT,
            'Mozilla/4.0 (compatible; MSIE 6.0; Windows NT 5.1; SV1; .NET CLR 1.0.3705; .NET CLR 1.1.4322)'
        );
```

As far as I can remember, the API isn't blocking any requests based on the user-agent, but if you're getting any requests denied due to the user-agent, do please let me know.

**Comment by rosell-dk** [comment]
I no longer get the 403 forbidden, so I have changed the User-Agent. This issue is done in this library. I however still need to implement deactivation of invalid keys in WebP Express (https://github.com/rosell-dk/webp-express/issues/326)

**Comment by rosell-dk** [comment]
@nosilver4u: 

It seems I get the same response for invalid and exceeded keys.
I get the following response when using a bogus key:

`{"error":"invalid","t":"exceeded"}`

**Comment by nosilver4u** [comment]
That is the correct response for a bogus (or expired) key, but for an exceeded key it should be this:
`{"error":"exceeded"}`

The extra t=exceeded is added for invalid keys, because older versions of EWWW IO didn't have a check for an "invalid" status.
.

**Comment by rosell-dk** [comment]
Released in WebP Express 0.17.0 and webp-convert 2.2.0
